Fix compilation on Ubuntu 22.04.

lmwnshn opened the pull request:
This PR fixes compilation errors on Ubuntu 22.04 caused by linkage issues (fix #4).

Please see the commit log for a quick summary of changes.

# Notes

1. I hope I preserved the right semantics with "what to extern" vs. "what to split into its own global variable", but you should probably check.
2. Additionally, `y.tab.c` and `y.tab.h` are build-time files that have been checked into source control. They changed because I am using a slightly newer version of Bison (you use 3.5.1, I use 3.8.2). Do you want them in the PR?
```
 code/tools/y.tab.c | 1114 ++++++++++++++++++++++++++++++++++++++++++++++++++++++--------------------------------------------------------------------
 code/tools/y.tab.h |  120 ++++++++------
 2 files changed, 554 insertions(+), 680 deletions(-)
```

Thanks!

lmwnshn committed:
[fix] Extern yydebug.

lmwnshn committed:
[fix] Multiple definitions of pDemoRecord.

lmwnshn committed:
[fix] Multiple definitions of pCountyRecord and pStateRecord.

lmwnshn committed:
[fix] Extern g_w_catalog_page.

lmwnshn committed:
[fix] Extern g_w_warehouse.

lmwnshn committed:
[fix] Extern g_w_web_site.

lmwnshn committed:
[fix] Explicitly rename and extern g_s_catalog_order_nItemIndex.

lmwnshn committed:
[fix] Explicitly rename and extern g_s_web_order_nItemIndex; extern g_s_web_order_lineitem.

lmwnshn committed:
[fix] Explicitly rename and extern g_s_purchase_nItemIndex.

lmwnshn commented:
@microsoft-github-policy-service agree

bailud commented on code/tools/QgenMain.c:
```diff
@@ -68,7 +68,7 @@ int g_nQueryNumber,
 
 
-int yydebug;
+extern int yydebug;
```
This change causes compilation failure on my side. Could you revert this change and see if it compiles on your side?

lmwnshn replied:
I tried reverting the change; without externing `yydebug`, I get a linker error when trying to compile `dsqgen`:

```
$ make
...
gcc -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -DYYDEBUG  -DLINUX -g -Wall  -o dsqgen address.o date.o decimal.o dist.o error_msg.o expr.o eval.o genrand.o grammar_support.o keywords.o list.o nulls.o print.o permute.o QgenMain.o query_handler.o r_params.o scaling.o StringBuffer.o substitution.o tdefs.o text.o tokenizer.o w_inventory.o y.tab.o release.o scd.o build_support.o parallel.o  utility.o -lm
/usr/bin/ld: y.tab.o:/home/wanshenl/git/dsb/code/tools/y.tab.c:1245: multiple definition of `yydebug'; QgenMain.o:/home/wanshenl/git/dsb/code/tools/QgenMain.c:71: first defined here
collect2: error: ld returned 1 exit status
make: *** [makefile:235: dsqgen] Error 1
```

My environment is the following:
```
$ lsb_release -cdr
Description:    Ubuntu 22.04 LTS
Release:        22.04
Codename:       jammy
$ uname -r
5.15.0-39-generic
$ gcc --version
gcc (Ubuntu 11.2.0-19ubuntu1) 11.2.0
$ bison --version
bison (GNU Bison) 3.8.2
$ flex --version
flex 2.6.4
```

Hmm. I will try to look for a workaround.

lmwnshn replied:
I found a workaround that involves specifying `-fcommon` as a `gcc` compiler flag. The details are in the commit message for the last commit (`42896f5`), essentially, `gcc-10` changed default behavior for whether the linker can merge global variables with the same name.

I wonder if you will have similar compilation errors for the other variables that I externed,
if so, please let me know and I will remove those `extern`s too.

I tested a bit further, just the addition of `-fcommon` is enough to make all of the initial code compile on my environment.
However, while `-fcommon` is safe to use for `yydebug` since there should only be one instance of that variable,
the other variables that I renamed and externed may not be safe to combine as they seem to be conceptually distinct.
For example, `pDemoRecord` was originally defined in two ways:

```
// w_customer_demographics.c
                int nMaxAttribute = distsize("gender") * distsize("education") * distsize("marital_status");
		int nKey = (int)get_rowcount(CUSTOMER_DEMOGRAPHICS);
		pDemoRecord = initializeCharAttributeKeyRecord(nMaxAttribute, nKey);

// w_household_demographics.c
                int nMaxAttribute = distsize("buy_potential") * distsize("dependent_count");
		int nKey = (int)get_rowcount(HOUSEHOLD_DEMOGRAPHICS);
		pDemoRecord = initializeCharAttributeKeyRecord(nMaxAttribute, nKey);
```

and it seems like we would not want to merge these definitions together.

So I think we would want to (1) specify `-fcommon` and (2) keep the variable renaming from this PR.

Thanks!

bailud replied:
> I tried reverting the change; without externing `yydebug`, I get a linker error when trying to compile `dsqgen`:
> 
> ```
> $ make
> ...
> gcc -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -DYYDEBUG  -DLINUX -g -Wall  -o dsqgen address.o date.o decimal.o dist.o error_msg.o expr.o eval.o genrand.o grammar_support.o keywords.o list.o nulls.o print.o permute.o QgenMain.o query_handler.o r_params.o scaling.o StringBuffer.o substitution.o tdefs.o text.o tokenizer.o w_inventory.o y.tab.o release.o scd.o build_support.o parallel.o  utility.o -lm
> /usr/bin/ld: y.tab.o:/<snip>/dsb/code/tools/y.tab.c:1245: multiple definition of `yydebug'; QgenMain.o:/<snip>/dsb/code/tools/QgenMain.c:71: first defined here
> collect2: error: ld returned 1 exit status
> make: *** [makefile:235: dsqgen] Error 1
> ```
> 
> My environment is the following:
> 
> ```
> $ lsb_release -cdr
> Description:    Ubuntu 22.04 LTS
> Release:        22.04
> Codename:       jammy
> $ uname -r
> 5.15.0-39-generic
> $ gcc --version
> gcc (Ubuntu 11.2.0-19ubuntu1) 11.2.0
> $ bison --version
> bison (GNU Bison) 3.8.2
> $ flex --version
> flex 2.6.4
> ```
> 
> Hmm. I will try to look for a workaround.

Hi Wan, could you check if yydebug is defined elsewhere in your local repo where the change is not included in the PR? Maybe check if the yydebug is defined in y.tab.c in your local repo, and try to revert that as well and see if it can compile without the linking error.

lmwnshn replied:
Hi Bailu,

As of commit 5a5c3bcaf5f501019203c2c498e33bc746431fed , `git status` returns a clean work tree. From there,

`yydebug` is already defined by DSB: https://github.com/microsoft/dsb/blob/main/code/tools/y.tab.c#L1117
And this definition is active because our CFLAGS include `-DYYDEBUG`, https://github.com/microsoft/dsb/blob/main/code/tools/makefile#L62

Even if I hack the makefile so that it does not regenerate `y.tab.c` (by no-op-ing https://github.com/microsoft/dsb/blob/main/code/tools/makefile#L283-L284),
it will have a link error.

Could you please check if your local environment somehow not including `-DYYDEBUG`?
e.g., maybe you are using an IDE that overrides CFLAGS. I am running directly from the terminal.

---

That said, the latest commit 42896f5647c2da79961d24f634843e603d699729 should have unified behavior across old and new GCC versions, does it still not compile?

arselzer replied:
Hi @lmwnshn 

Thank you for providing the fix!

I also ran into the same problem when trying to run the makefile.

Using your branch at the latest commit, I was able to compile everything without errors.
In case it still helps, the environment I used was the following:

* Ubuntu 22.04.3 LTS
* 6.2.0-33-generic kernel
* gcc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
* bison (GNU Bison) 3.8.2
* flex 2.6.4

lmwnshn replied:
@arselzer Good to know, thanks for the info. I will link from the issue in case anyone else finds it helpful.

EvagelineFEI replied:
Hi @lmwnshn，
大佬真是救大命了What you do really helps me!! Thanks capybara

lmwnshn committed:
[revert] Unextern yydebug.

lmwnshn committed:
[fix] Add -fcommon to makefile.

gcc-10 changed the default from -fcommon to -fno-common.

Quoting the manual,

> In C code, this option controls the placement of global variables defined
> without an initializer, known as tentative definitions in the C standard.
> Tentative definitions are distinct from declarations of a variable with
> the extern keyword, which do not allocate storage.

> The default is -fno-common, which specifies that the compiler places
> uninitialized global variables in the BSS section of the object file.
> This inhibits the merging of tentative definitions by the linker so
> you get a multiple-definition error if the same variable is accidentally
> defined in more than one compilation unit.

> The -fcommon places uninitialized global variables in a common block.
> This allows the linker to resolve all tentative definitions of the same
> variable in different compilation units to the same object, or to a
> non-tentative definition. This behavior is inconsistent with C++, and on
> many targets implies a speed and code size penalty on global variable
> references. It is mainly useful to enable legacy code to link without errors.

In Bison's case, yydebug is generated as a tentative definition in y.tab.c.

> /* Nonzero means print parse trace. It is left uninitialized so that
>    multiple parsers can coexist. */
> int yydebug;

And therefore adding the -fcommon flag
1. maintains the same behavior between gcc versions pre-10 and 10+, and
2. allows for `yydebug` to not be externed, the linker will figure it out.
